lib/posix/pthread.c

```diff
@@ -436,24 +437,28 @@ int pthread_join(pthread_t thread, void **status)
 		return ESRCH;
 	}
 
-	pthread_mutex_lock(&pthread->state_lock);
-
-	if (pthread->state == PTHREAD_JOINABLE) {
-		pthread_cond_wait(&pthread->state_cond, &pthread->state_lock);
+	if (pthread->state == PTHREAD_DETACHED) {
+		return EINVAL;
 	}
 
+	k_thread_join(&pthread->thread, K_FOREVER);
+
+	pthread_mutex_lock(&pthread->state_lock);
```

Comment:
This part is still kind of bothering me..

Reply:
Could you use atomics for the state instead of a mutex?

Reply:
Alternatively, don't free the mutex ever. Use a static initializer for the thread pool. Or use a single mutex for all threads.

Reply:
Should be fixed with the latest rev. And actually pretty tight, I hope.